ysh/funcs_builtin.py

```diff
@@ -21,121 +22,195 @@
 
 
 def SetGlobalFunc(mem, name, func):
-    # type: (state.Mem, str, Union[Callable, type]) -> None
+    # type: (state.Mem, str, Union[Callable, type, vm._Func]) -> None
     """Used by bin/oil.py to set split(), etc."""
-    assert callable(func), func
-
     # TODO: Fix this location info
     left = lexer.DummyToken(Id.Undefined_Tok, '')
-    mem.SetValue(sh_lhs_expr.Name(left, name), value.Obj(func),
-                 scope_e.GlobalOnly)
+    if isinstance(func, vm._Func):
+        mem.SetValue(sh_lhs_expr.Name(left, name), value.Func(func),
+                     scope_e.GlobalOnly)
+    else:
+        assert callable(func), func
+        mem.SetValue(sh_lhs_expr.Name(left, name), value.Obj(func),
+                     scope_e.GlobalOnly)
 
 
-def _Join(array, delim=''):
-    """Func join(items Array[Str]) Str ..."""
-    # default is not ' '?
-    return delim.join(array)
+class _Join(vm._Func):
 
+    def __init__(self):
+        # type: () -> None
+        vm._Func.__init__(self)
 
-def _Maybe(obj):
-    """Func join(items Array[Str]) Str ..."""
-    if obj is None:
-        return []
+    def Run(self, pos_args, named_args):
+        # type: (List[value_t], Dict[str, value_t]) -> value_t
+        assert len(pos_args) == 2
 
-    # TODO: Need proper span IDs
-    if not isinstance(obj, str):
-        raise error.Expr('maybe() passed arg of invalid type %r' %
-                         obj.__class__.__name__)
+        # XXX: python's join() accepts dict keys, strings, and other iterables.
```

Comment:
Yeah this is a good question ... We don't have any concept of "iterators"

So it may be that users need to call `join(d->keys(), '')` and so forth

Not `''.join(d)` like in Python, which is shorter, but arguably a little implicit and hard to learn